lib/galaxy/workflow/scheduling_manager.py

```diff
@@ -329,6 +336,23 @@ def __schedule(self, workflow_scheduler_id, workflow_scheduler):
     def __attempt_schedule(self, invocation_id, workflow_scheduler):
         with self.app.model.context() as session:
             workflow_invocation = session.get(model.WorkflowInvocation, invocation_id)
+            if workflow_invocation.state == workflow_invocation.states.REQUIRES_MATERIALIZATION:
+                hdas_to_materialize = workflow_invocation.inputs_requiring_materialization()
+                for hda in hdas_to_materialize:
+                    user = RequestUser(user_id=workflow_invocation.history.user_id)
+                    task_request = MaterializeDatasetInstanceTaskRequest(
+                        user=user,
+                        history_id=workflow_invocation.history.id,
+                        source="hda",
+                        content=hda.id,
+                    )
+                    self.app.hda_manager.materialize(task_request, in_place=True)
```

Comment:
If you return a celery task here you could create a chord and let a celery task change the state to READY (or FAILED) on success/failure, and set the state here to something new like `AWAITING_MATERIALIZATION` or something like that.

Reply:
Can again do this in a follow up

Reply:
One thing to address before merging though: Will we keep on materializing indefinitely if materializing fails ? We may need an exception handler here so it doesn't get picked up again in the next scheduling round.

Reply:
That's an easy test - especially with the hash validation implemented now. I'll give it a shot.

Reply:
Good new and bad news. The good news is the materialize code is pretty robust it seems 😅 , the bad news - I don't know how to test it then 😆.

Bad hashes result in dataset errors and the jobs just pausing out.

```
| 2 - NCA1IDY= (HID - NAME) 
INFO:     127.0.0.1:56746 - "GET /api/histories/adb5f5c93f827949/contents/529fd61ab1c6cc36 HTTP/1.1" 200 OK
| Dataset State:
|  error
| Dataset Blurb:
|  *Dataset blurb was empty.*
| Dataset Info:
|  Failed to materialize deferred dataset with exception: Failed to validate path with [HashFunctionNameEnum.md5] - expected [abadmd5sumhash] got [ad0f811416f7ed2deb9122007d649fb0]
| Peek:
|  *Peek unavailable.*
INFO:     127.0.0.1:56747 - "GET /api/histories/adb5f5c93f827949/contents/529fd61ab1c6cc36/provenance HTTP/1.1" 200 OK
| Dataset Job Standard Output:
|  *Standard output was empty.*
| Dataset Job Standard Error:
|  *Standard error was empty.*
|
--------------------------------------
| 3 - Concatenate datasets on data 2 and data 1 (HID - NAME) 
galaxy.model WARNING 2024-09-27 14:22:44,912 [pN:main,p:82902,tN:AnyIO worker thread] Failed to determine file name for dataset 3
INFO:     127.0.0.1:56748 - "GET /api/histories/adb5f5c93f827949/contents/d9abeb98649a6a7e HTTP/1.1" 200 OK
| Dataset State:
|  paused
| Dataset Blurb:
|  queued
| Dataset Info:
|  Input dataset 'MSAyIDM=' is in error state, Input dataset 'NCA1IDY=' is in error state. To resume this job fix the input dataset(s).
| Peek:
|  *Peek unavailable.*
```

Bad URIs also provide nice user errors and just finish the scheduling.

```
| 1 - work (HID - NAME) 
INFO:     127.0.0.1:57308 - "GET /api/histories/adb5f5c93f827949/contents/adb5f5c93f827949 HTTP/1.1" 200 OK
| Dataset State:
|  error
| Dataset Blurb:
|  *Dataset blurb was empty.*
| Dataset Info:
|  Failed to materialize deferred dataset with exception: Could not find handler for URI [gxfiles://thisurl/doesnt/work]
| Peek:
|  *Peek unavailable.*
INFO:     127.0.0.1:57309 - "GET /api/histories/adb5f5c93f827949/contents/adb5f5c93f827949/provenance HTTP/1.1" 200 OK
| Dataset Job Standard Output:
|  *Standard output was empty.*
| Dataset Job Standard Error:
|  *Standard error was empty.*
|
--------------------------------------
| 2 - work (HID - NAME) 
INFO:     127.0.0.1:57310 - "GET /api/histories/adb5f5c93f827949/contents/529fd61ab1c6cc36 HTTP/1.1" 200 OK
| Dataset State:
|  error
| Dataset Blurb:
|  *Dataset blurb was empty.*
| Dataset Info:
|  Failed to materialize deferred dataset with exception: Could not find handler for URI [gxfiles://thisurl/doesnt/work]
| Peek:
|  *Peek unavailable.*
INFO:     127.0.0.1:57311 - "GET /api/histories/adb5f5c93f827949/contents/529fd61ab1c6cc36/provenance HTTP/1.1" 200 OK
| Dataset Job Standard Output:
|  *Standard output was empty.*
| Dataset Job Standard Error:
|  *Standard error was empty.*
|
--------------------------------------
| 3 - Concatenate datasets on data 2 and data 1 (HID - NAME) 
galaxy.model WARNING 2024-09-27 14:30:11,372 [pN:main,p:83480,tN:AnyIO worker thread] Failed to determine file name for dataset 3
INFO:     127.0.0.1:57312 - "GET /api/histories/adb5f5c93f827949/contents/d9abeb98649a6a7e HTTP/1.1" 200 OK
| Dataset State:
|  paused
| Dataset Blurb:
|  queued
| Dataset Info:
|  Input dataset 'work' is in error state, Input dataset 'work' is in error state. To resume this job fix the input dataset(s).
| Peek:
|  *Peek unavailable.*
INFO:     127.0.0.1:57313 - "GET /api/histories/adb5f5c93f827949/contents/d9abeb98649a6a7e/provenance HTTP/1.1" 200 OK
| Dataset Job Standard Output:
|  *Standard output was empty.*
| Dataset Job Standard Error:
|  *Standard error was empty.*
```

Reply:
> Bad hashes result in dataset errors and the jobs just pausing out.

And we're definitely not attempting the schedule and materialize the invocation again ? At the end of the test the invocation is in failed state ? The "regular" error handling is downstream for invocations is in https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/workflow/run.py#L100, that's why I am a little worried about this.

Reply:
Ah, I understand now, since we never raise any exceptions we just go on scheduling and that then fails with the messages above. That makes sense, even though I guess I would have slightly preferred to explicitly say that input materialization failed. That's fine.

Reply:
I've added a bunch of tests around this - a positive validation and two different ways to fail during materialization. I also added untested exception handling in case some else goes wrong during materialization. A more formal failure in this case makes some sense to me - it seems like a fine UI enhancement we can just add if this feature gets used though I think. I'm always anxious about polishing the unused and experimental.